### ISSUE TITLE
robots.txt rules turned into overly greedy regex

### ISSUE THREAD

**Issue by niels** [Norconex/collector-http#200]
Given the below robots.txt, a URL such as http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html is incorrectly rejected.

```
User-agent: Mediapartners-Google
Disallow:

User-agent: Baiduspider
Disallow: /

User-agent: *
Disallow: /*addtofav$
Disallow: /errorlog/
Disallow: /Mail/
Disallow: /WebResource.axd
Disallow: /+/categorypath
Disallow: /%23all
Disallow: /*.pdf
Disallow: /53597944
Disallow: /20866434
Disallow: /calculator
Disallow: *ctl=
Disallow: *?prodid=
Disallow: *ctgn=
Disallow: /*mascusproduct
Disallow: /productCard3print.aspx
Disallow: /tellafriend.aspx
Sitemap: http://www.mascus.com/sitemap_index_com.xml
```

I believe this is due to the `Disallow: /*.pdf` stance in particular which [StandardRobotsTxtProvider.RobotData.buildURLFilter()](https://github.com/Norconex/collector-http/blob/d2c598085437a91982d77957a55c1514ae9a1e1e/norconex-collector-http/src/main/java/com/norconex/collector/http/robot/impl/StandardRobotsTxtProvider.java#L239) seems to turn into the regex `/.*.pdf?` when in fact it should be `/.*\.pdf`. As a result, all URLs containing any character followed by the sequence "pd" get rejected.

Note that (a) the `.` should be escaped (i.e. taken literally) as otherwise it would match *any* character and (b) the final `?` should not be added as this would match on `pd` as well as `pdf`, the former of which is not intended. This is as per [Google's interpretation](https://support.google.com/webmasters/answer/6062596?hl=en):

![screenshot-support google com 2015-12-09 13-49-56](https://cloud.githubusercontent.com/assets/16709/11685772/d13b33dc-9e7b-11e5-8220-82baeb36732a.png)

Sample config:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<httpcollector id="test-collector">
  <crawlers>
    <crawler id="test-crawler">
      <sitemapResolverFactory class="com.norconex.collector.http.sitemap.impl.StandardSitemapResolverFactory" ignore="true" lenient="true" />

      <startURLs>
        <sitemap>http://www.mascus.com/googleXML/sitemaps/com/com_agriculture_productcard_1.xml</sitemap>
      </startURLs>
    </crawler>
  </crawlers>
</httpcollector>
```

Sample rejection:

```
test-crawler: 2015-12-09 13:31:49 DEBUG - ACCEPTED document reference (robots.txt). Reference=http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html Filter=Robots.txt (Disallow:/*addtofav$)
test-crawler: 2015-12-09 13:31:49 DEBUG - ACCEPTED document reference (robots.txt). Reference=http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html Filter=Robots.txt (Disallow:/errorlog/)
test-crawler: 2015-12-09 13:31:49 DEBUG - ACCEPTED document reference (robots.txt). Reference=http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html Filter=Robots.txt (Disallow:/Mail/)
test-crawler: 2015-12-09 13:31:49 DEBUG - ACCEPTED document reference (robots.txt). Reference=http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html Filter=Robots.txt (Disallow:/WebResource.axd)
test-crawler: 2015-12-09 13:31:49 DEBUG - ACCEPTED document reference (robots.txt). Reference=http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html Filter=Robots.txt (Disallow:/+/categorypath)
test-crawler: 2015-12-09 13:31:49 DEBUG - ACCEPTED document reference (robots.txt). Reference=http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html Filter=Robots.txt (Disallow:/%23all)
test-crawler: 2015-12-09 13:31:49 DEBUG - REJECTED document reference  (robots.txt). Reference=http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html Filter=Robots.txt (Disallow:/*.pdf)
test-crawler: 2015-12-09 13:31:49 INFO -           REJECTED_FILTER: http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html (Subject: Robots.txt (Disallow:/*.pdf))
test-crawler: 2015-12-09 13:31:49 INFO -       REJECTED_ROBOTS_TXT: http://www.mascus.com/agriculture/used-combination-drills/tume-kl-2500/ckckapdt.html (Subject: RobotsTxt[filters={Robots.txt (Disallow:/*addtofav$),Robots.txt (Disallow:/errorlog/),Robots.txt (Disallow:/Mail/),Robots.txt (Disallow:/WebResource.axd),Robots.txt (Disallow:/+/categorypath),Robots.txt (Disallow:/%23all),Robots.txt (Disallow:/*.pdf),Robots.txt (Disallow:/53597944),Robots.txt (Disallow:/20866434),Robots.txt (Disallow:/calculator),Robots.txt (Disallow:*ctl=),Robots.txt (Disallow:*?prodid=),Robots.txt (Disallow:*ctgn=),Robots.txt (Disallow:/*mascusproduct),Robots.txt (Disallow:/productCard3print.aspx),Robots.txt (Disallow:/tellafriend.aspx)},crawlDelay=-1.0,sitemapLocations={http://www.mascus.com/sitemap_index_com.xml}])
```

**Comment by essiembre** [comment]
Could reproduce. Definitely an error.  Will fix.

**Comment by essiembre** [comment]
Fixed in the latest [snapshot release](https://www.norconex.com/collectors/collector-http/download). Thanks for catching this.  Please test.

**Comment by niels** [comment]
Thanks for the fix! I can confirm that the latest snapshot correctly crawls the previously erroneously rejected URLs mentioned above.

Regarding the implementation, perhaps [Pattern.quote()](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html#quote-java.lang.String-) would have been a more robust choice?

**Comment by essiembre** [comment]
Thanks for confirming! Yeah, good point about quote().  I thought about it but since we need the replaced .* to remain a regex within an otherwise literal string, I figured it would not work and/or be simpler to do without it.  I could have reduced the many "replace" lines into a single expression though.  I think it's one of these I was just happy to get working fine. :-)  It's funny you caught that given I committed it with hesitation (not as "clean" as I usually like).  I guess there is just no escaping for me!  Feel free to submit a patch if you are OK to test it again afterwards.  Otherwise, I may revise it once it has haunted me too much. :-)

**Comment by niels** [comment]
Ha.. sorry for catching you out like that :) I actually think the code reads plenty clean as-is. My concern is more along the lines of "Did we forget to escape something we should have escaped?". In particular, regular expression engines tend to have slight differences between implementation languages and I am too inexperienced with Java's implementation to feel perfectly safe with escaping manually vs. using built-in (and thus presumably secure) functionality.

I totally get your point regarding the inelegance of first escaping and then "un-escaping". My personal rule of thumb would be: if there's more to escape then un-escape, globally do a wholesale escape and then individually un-escape as needed vs. selectively escape so that there's nothing to un-escape.

To that end, I will indeed submit a patch later this week and then leave it up to you which implementation you feel more comfortable with going forward.

A couple of weeks ago, I also modified the elasticsearch committer to use ES's REST interface, so I am trying to get more involved in development of the ecosystem. I developed this as a separate committer so that users can choose for themselves whether to use the Node client (of the existing committer) or the REST client (of my new committer).

As I haven't yet gotten around to writing a test suite, I was thus far too hesitant to "officially submit" the new committer to you. But seeing how I am critiquing *your* "secret" commits, it is only fair to expose my own.. So here it is, if you are interested in taking a peek: https://github.com/herimedia/norconex-committer-elasticsearch-rest.
